File: 1393-MaximumValueOfKCoinsFromPiles/1393-MaximumValueOfKCoinsFromPiles.py
# Last updated: 4/29/2026, 10:24:42 AM
"""
 remain, 3, 3
    max(piles[0] + dp(remain - 1, 2, 3), piles[1] + dp(remain-1, 3, 2))
    if remain == 
"""
class Solution:
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:
        @cache
        def dp(i, remain):
            if i == len(piles) or remain == 0:
                return 0
            take = 0
            pile_len = len(piles[i])
            curr = 0
            for coin_num in range(1, min(remain, pile_len) + 1):
                curr += piles[i][coin_num - 1]
                take = max(take, curr + dp(i+1, remain-coin_num))
            return max(dp(i+1, remain), take)
        
        return dp(0, k)

File: 1393-MaximumValueOfKCoinsFromPiles/1393-MaximumValueOfKCoinsFromPiles.js
// Last updated: 4/29/2026, 10:24:37 AM
/**
 * @param {number[][]} piles
 * @param {number} k
 * @return {number}
 */
var maxValueOfCoins = function(piles, k) {
    /*
    remain, 3, 3
    max(piles[0] + dp(remain - 1, 2, 3), piles[1] + dp(remain-1, 3, 2))
    if remain == 
    */
    const n = piles.length;
    const memo = new Array(n+1).fill(null).map(() => new Array(k+1).fill(-1));
    const dp = (i, remain) => {
        if(i === n || remain === 0) {
            return 0;
        }
        if(memo[i][remain] !== -1) {
            return memo[i][remain];
        }
        const skip = dp(i+1, remain);
        let take = 0;
        const maxCoin = Math.min(remain, piles[i].length);
        let curr = 0;
        for(let j = 1; j <= maxCoin; j++) {
            curr += piles[i][j-1];
            take = Math.max(take, curr + dp(i+1, remain-j))
        }
        memo[i][remain] = Math.max(skip, take);
        return memo[i][remain];
    }
    
    return dp(0, k);
};